File: ruby/protein-translation/protein_translation.rb
module Translation
  class Error < StandardError; end
  class InvalidCodonError < Error; end

  STOP = "STOP"

  DICTIONARY = {
    %w{UUA UUG}         => 'Leucine',
    %w{AUG}             => 'Methionine',
    %w{UUC UUU}         => 'Phenylalanine',
    %w{UCU UCC UCA UCG} => 'Serine',
    %w{UAU UAC}         => 'Tyrosine',
    %w{UGU UGC}         => 'Cysteine',
    %w{UGG}             => 'Tryptophan',
    %w{UAA UAG UGA}     => STOP,
  }.map.with_object({}) { |(ks,v),h| ks.each { |k| h[k] = v } }

  def self.of_codon(codon)
    DICTIONARY[codon]
  end

  def self.of_rna(strand)
    strand.scan(/.../).each.with_object([]) do |codon, proteins|
      case protein = Translation.of_codon(codon)
      when STOP then return proteins
      when nil then raise InvalidCodonError
      else proteins << protein
      end
    end
  end
end

File: haskell/protein-translation/src/ProteinTranslation.hs
{-# LANGUAGE OverloadedStrings #-}

module ProteinTranslation(proteins) where

type Codon = String
type Protein = String
type Sequence = String

proteins :: Sequence -> Maybe [Protein]
proteins sequence = takeWhile (/= "STOP") <$> translateSequence sequence

translateSequence :: Sequence -> Maybe [Protein]
translateSequence [] = Just []
translateSequence sequence = (:) <$> translateCodon codon <*> translateSequence remainder
  where (codon, remainder) = splitAt 3 sequence

translateCodon :: Codon -> Maybe Protein
translateCodon c
  | elem c ["AUG"]                      = Just "Methionine"
  | elem c ["UUU", "UUC"]               = Just "Phenylalanine"
  | elem c ["UUA", "UUG"]               = Just "Leucine"
  | elem c ["UCU", "UCC", "UCA", "UCG"] = Just "Serine"
  | elem c ["UAU", "UAC"]               = Just "Tyrosine"
  | elem c ["UGU", "UGC"]               = Just "Cysteine"
  | elem c ["UGG"]                      = Just "Tryptophan"
  | elem c ["UAA", "UAG", "UGA"]        = Just "STOP"
  | otherwise                           = Nothing
